control-plane/agents/src/bin/ha/cluster/etcd.rs

```diff
@@ -0,0 +1,99 @@
+    },
+};
+use futures::lock::Mutex;
+use std::{convert::TryFrom, sync::Arc, time::Duration};
+use tracing::debug;
+
+#[derive(Debug, Clone)]
+pub struct EtcdStore {
+    store: Arc<Mutex<Etcd>>,
+    timeout: Duration,
+}
+
+impl EtcdStore {
+    pub async fn new(endpoint: String, timeout: Duration) -> Result<Self, StoreError> {
+        match tokio::time::timeout(timeout, async { Etcd::new(&endpoint).await }).await {
+            Ok(v) => {
+                let store = v?;
+                Ok(Self {
+                    store: Arc::new(Mutex::new(store)),
+                    timeout,
+                })
+            }
+            Err(_) => Err(StoreError::Timeout {
+                operation: "connect".to_string(),
+                timeout,
+            }),
+        }
+    }
+
+    /// Serialized write to the persistent store
+    pub async fn store_obj<O: StorableObject>(&self, object: &O) -> Result<(), anyhow::Error> {
+        let mut store = self.store.lock().await;
+        match tokio::time::timeout(self.timeout, async move { store.put_obj(object).await }).await {
+            Ok(result) => result.map_err(Into::into),
+            Err(_) => Err(StoreError::Timeout {
+                operation: "Put".to_string(),
+                timeout: self.timeout,
+            }
+            .into()),
+        }
+    }
+
+    /// Serialized read from the persistent store
+    pub async fn load_obj<O: StorableObject>(&self, key: &O::Key) -> Result<O, anyhow::Error> {
+        let mut store = self.store.lock().await;
+        match tokio::time::timeout(self.timeout, async move { store.get_obj(key).await }).await {
+            Ok(obj) => Ok(obj?),
+            Err(_) => Err(StoreError::Timeout {
+                operation: "Get".to_string(),
+                timeout: self.timeout,
+            }
+            .into()),
+        }
+    }
+
+    /// get incomplete requests stored in Etcd
+    /// request with error or path published is considered a complete request
+    pub async fn fetch_incomplete_requests(&self) -> Result<Vec<SwitchOverRequest>, anyhow::Error> {
+        let mut store = self.store.lock().await;
+        let key = key_prefix_obj(StorableObjectType::SwitchOver);
+        let store_entries = store.get_values_prefix(&key).await?;
+
+        debug!("fetched entries from etcd");
+
+        let store_values: Vec<serde_json::Value> =
+            store_entries.iter().map(|v| v.1.clone()).collect();
+
+        let entries: Vec<Result<SwitchOverSpec, serde_json::Error>> = store_values
+            .iter()
+            .map(|v| serde_json::from_value(v.clone()))
+            .collect();
+
+        let mut ee = vec![];
+        for spec in entries {
+            match spec {
+                Ok(s) => {
+                    if !s.is_completed() {
+                        let req = SwitchOverRequest::try_from(&s)?;
+                        ee.push(req)
+                    }
+                }
+                Err(e) => {
```

Comment:
Do we want to error out if any fails? If so you can make use of the collect above to tidy this up:
```rust
        let entries = store_entries
            .into_iter()
            .map(|v| v.1)
            .map(serde_json::from_value)
            .collect::<Result<Vec<SwitchOverSpec>, serde_json::Error>>()?;
```